File: cxyz/lib/qgate/cry.cpp
/*
 * Author: Hanyu Wang
 * Created time: 2024-03-31 14:45:41
 * Last Modified by: Hanyu Wang
 * Last Modified time: 2024-04-02 14:08:42
 */

#include "qstate.hpp"
#include "qgate.hpp"

#include <cmath>

namespace xyz
{
QState CRY::operator()(const QState& state, const bool reverse) const
{
    auto _theta = reverse? -theta : theta;
    QState new_state;
    for (const auto& [index, weight] : state.index_to_weight)
    {
        if ((bool)((index >> ctrl) & 1) != phase)
        {
            new_state.index_to_weight[index] = weight;
            continue;
        }
        if (new_state.index_to_weight.find(index) == new_state.index_to_weight.end())
            new_state.index_to_weight[index] = 0;
        new_state.index_to_weight[index] += std::cos(_theta / 2) * weight;
        uint32_t new_index = index ^ (1 << target_qubit);
        new_state.index_to_weight[new_index] += std::sin(_theta / 2) *
        ((index & (1 << target_qubit))? -weight : weight);
    }
    for (auto it = new_state.index_to_weight.begin(); it != new_state.index_to_weight.end();)
    {
        if (std::abs(it->second) < QState::eps)
            it = new_state.index_to_weight.erase(it);
        else
            ++it;
    }
    new_state.n_bits = state.n_bits;
    return new_state;
}
uint32_t CRY::get_cost() const
{
    return 2;
}
std::string CRY::to_string() const
{
    std::string ctrl_str = Controlled::to_string();
    std::string ry_str = RY::to_string();
    return ctrl_str + ry_str;
}
} // namespace xyz

File: xyz/circuit/gate/cry.py
#!/usr/bin/env python
# -*- encoding=utf8 -*-

"""
Author: Hanyu Wang
Created time: 2023-06-22 13:11:53
Last Modified by: Hanyu Wang
Last Modified time: 2023-06-22 23:32:05
"""

import numpy as np

from .base import BasicGate, ControlledGate, QBit, QGateType, RotationGate
from .qstate import QState


class CRY(RotationGate, BasicGate, ControlledGate):
    """Controlled Y Rotation."""

    def __init__(
        self, theta: float, control_qubit: QBit, phase: int, target_qubit: QBit
    ) -> None:
        BasicGate.__init__(self, QGateType.CRY, target_qubit)
        RotationGate.__init__(self, theta)
        ControlledGate.__init__(self, control_qubit, phase)

    def __str__(self) -> str:
        control_str = "" if self.phase == 1 else "~"
        return f"C({control_str}{self.control_qubit})RY({self.target_qubit}:{self.theta:0.02f})"

    def get_cnot_cost(self) -> int:
        """Returns the cost of the cost of the gate."""
        return 2

    def conjugate(self) -> "CRY":
        return CRY(-self.theta, self.control_qubit, self.phase, self.target_qubit)

    def apply(self, qstate: QState) -> QState:
        index_to_weight = {idx: 0 for idx in qstate.index_set}
        for idx, weight in qstate.index_to_weight.items():
            # no rotation
            if not self.is_enabled(idx):
                index_to_weight[idx] = qstate.index_to_weight[idx]
                continue
            rdx = idx ^ (1 << self.target_qubit.index)
            if rdx not in qstate.index_set:
                index_to_weight[rdx] = 0

            if (idx >> self.target_qubit.index) & 1 == 0:
                index_to_weight[idx] += weight * np.cos(self.theta / 2)
                index_to_weight[rdx] += weight * np.sin(self.theta / 2)
            else:
                index_to_weight[idx] += weight * np.cos(self.theta / 2)
                index_to_weight[rdx] -= weight * np.sin(self.theta / 2)
        return QState(index_to_weight, qstate.num_qubits)
